Testing

yuyuranium opened the pull request:
I can reproduce the problem encountered in #6  so I wrote some test to test both BRAM and SuperBRAM. However, no matter how I tried, the result seems to be the same: The initial values in both BRAM and SuperBRAM are all shifted by one word. See the program output below:

![image](https://user-images.githubusercontent.com/79467307/169706196-d8d1d6e3-a436-4569-a014-e73d22f52d64.png)

The output is generated by the function `dump_bram` in `main.c`.

I also did some test on writing data to brams and reading it back by the codes I commented below. The result was weird:

![image](https://user-images.githubusercontent.com/79467307/169706271-c607c489-6223-4e1d-907d-b704126d364a.png)

The main program should write `0x87` to word offset 0, 5, and 8 of both brams (But it's offset 1, 6, and 9 that become `0x87`). From the picture we can tell that all writing operations are shifted by one word also. Besides, the offset 0 of both brams were also be written.

BTW, the good news is that `codegen` function is working fine.

yuyuranium committed:
Copy Willy's main.c and add dump_bram function

yuyuranium committed:
Include xil_io.h in controller.h

yuyuranium committed:
Add tests for BRAM read/write

yuyuranium commented on src/main.c:
```diff
@@ -0,0 +1,111 @@
+	dump_bram();
+	//exec(XPAR_CONTROLLER_0_S00_AXI_BASEADDR,
+		// codegen(0, 0b0000101, 0b10001, 3 + 1, 2 + 1, 0 + 1)); // alumode = 0000, opmode = 0000101, inmode = 10001, bram1wAddr = 4*3, bram1rAddr = 4*2, bram0rAddr = 4*0;
+	Xil_Out32(XPAR_AXI_BRAM_CTRL_1_S_AXI_BASEADDR, 0x87);
+	Xil_Out32(XPAR_AXI_BRAM_CTRL_1_S_AXI_BASEADDR + 20, 0x87);
+	Xil_Out32(XPAR_AXI_BRAM_CTRL_0_S_AXI_BASEADDR, 0x87);
+	Xil_Out32(XPAR_AXI_BRAM_CTRL_0_S_AXI_BASEADDR + 20, 0x87);
+	Xil_Out32(XPAR_AXI_BRAM_CTRL_1_S_AXI_BASEADDR + 32, 0x87);
+	Xil_Out32(XPAR_AXI_BRAM_CTRL_0_S_AXI_BASEADDR + 32, 0x87);
```
I used these codes to test reading/writing the brams.

yuyuranium committed:
Update test for bram output registers

yuyuranium commented:
## Update
I modified the `dump_bram` function to read a single word multiple times and output the final result. Now it looks all good!

- Initial value
![image](https://user-images.githubusercontent.com/79467307/169731924-f05bb060-2b68-4bab-a679-c08e81dfe004.png)

- After writing `0x87` to 0, 5, 8 word offsets
![image](https://user-images.githubusercontent.com/79467307/169732112-80689f45-4c28-4a5a-8414-59b0f5e618c0.png)

yuyuranium commented on src/main.c:
```diff
@@ -13,9 +13,12 @@
-				i, Xil_In32(XPAR_AXI_BRAM_CTRL_0_S_AXI_BASEADDR + 4 * i),
-				i, Xil_In32(XPAR_AXI_BRAM_CTRL_1_S_AXI_BASEADDR + 4 * i));
+		u32 a, b;
+		for (int j = 0; j < 10; ++j) {
+			a = Xil_In32(XPAR_AXI_BRAM_CTRL_0_S_AXI_BASEADDR + 4 * i);
+			b = Xil_In32(XPAR_AXI_BRAM_CTRL_1_S_AXI_BASEADDR + 4 * i);
+		}
```
I used this for loop to read the same entry multiple times (exactly 10 times) to get a steady result.

yuyuranium commented:
## Update
Good news! Thanks to #9, the brams work perfectly!

WillyChennnn commented:
Finally,
thanks for testing！

yuyuranium commented:
Bad news is that the computation seems not to be carried out. I am checking if the output `P` from DSP has value or not by instanciate an AXI_GPIO so that the PS can read the value of `P` and print it out on the terminal. QQ

yuyuranium committed:
Remove bram testing code and add DSP testing code

yuyuranium commented:
I have test the value of `P` via AXI_GPIO. However, the value before and after executing the command is always 0.

![image](https://user-images.githubusercontent.com/79467307/169813719-7d195e0c-50ae-49b8-992d-d373e82f67dc.png)

yuyuranium commented on IPs/controller_1.0/drivers/controller_v1_0/src/controller.c:
```diff
@@ -10,9 +10,16 @@ void exec(UINTPTR baseAddr, u32 ins)
 
     // Write start signal to control register 0
     CONTROLLER_mWriteReg(baseAddr, 0, 1);
+#include "xparameters.h"
+    printf("Before: %lx\n", Xil_In32(XPAR_AXI_GPIO_0_BASEADDR));
 
     // Busy waiting for execution
     while (CONTROLLER_mReadReg(baseAddr, 0));
+    printf("After: %lx\n", Xil_In32(XPAR_AXI_GPIO_0_BASEADDR));
+    printf("After: %lx\n", Xil_In32(XPAR_AXI_GPIO_0_BASEADDR));
+    printf("After: %lx\n", Xil_In32(XPAR_AXI_GPIO_0_BASEADDR));
+    printf("After: %lx\n", Xil_In32(XPAR_AXI_GPIO_0_BASEADDR));
+    printf("After: %lx\n", Xil_In32(XPAR_AXI_GPIO_0_BASEADDR));
```
I used these code to read the value from AXI_GPIO, which is added to the block design by me to read the output of DSP directly.

yuyuranium committed:
Revise bram enb signals in controller

yuyuranium committed:
Update controller IP

yuyuranium committed:
Modify controller to start-valid-clear protocal

yuyuranium committed:
Revise main.c

Move the codegen and exec function to main.c. Now the invokation of the
execution is set via GPIO. We also have to reset the enable signal after
valid is pulled up.

yuyuranium commented:
### Update
Good news! The system seems to work. I revise the controller and how it communicates with PS. Now PS has to use AXI GPIO to invoke the execution of an instruction by
1. Set the instruction to controller 
2. Write enable signal to controller
3. Wait for valid signal from controller
4. Reset enable signal to controller

Here is the result of executing the first instruction:
- Before
![image](https://user-images.githubusercontent.com/79467307/169879556-5908db9a-1136-4815-814f-b13b3f625b40.png)

- After
![image](https://user-images.githubusercontent.com/79467307/169879597-bcea72c1-e106-41e4-8b35-1aebd22eaf41.png)

We can see that after executiong SuperBRAM[3] is written with correct value.

yuyuranium commented:
@WillyChennnn would you help me translate the pseudo code into a series of `codegen` function calls?

For example,
```
BRAM1[3] <= BRAM0[0] * BRAM1[2];
```
becomes
```
codegen(0b0000, 0b0000101, 0b10001, 3, 2, 0)
```

You may write the result directly in the comment section, thanks!

WillyChennnn commented:
Ok, but may I complete this tomorrow?
I will have an interview for internship in tomorrow morning, so I want to spend some time preparing it😥

yuyuranium commented:
Oh! I noticed that you have done so in the comments in `main.c`!

yuyuranium committed:
Update main.c with all the instruction

yuyuranium commented:
### Update
Thanks to @WillyChennnn's dedication, our system finally works perfectly. Here is the execution result, which is totally identical to TA's answer.

![image](https://user-images.githubusercontent.com/79467307/169884326-e2073977-00c5-4fac-8b37-1d4cf7c1dddd.png)
![image](https://user-images.githubusercontent.com/79467307/169884358-c110c71f-1174-436b-8a5b-48764b1e84eb.png)

I think this PR is ready for review. Please let me know if you have comments. Thanks a lot!

WillyChennnn commented:
Congratulations
Thanks for big hardworking!

yuyuranium committed:
Improve coding style of main.c

yuyuranium committed:
Add vitis project

yuyuranium committed:
Remove useless controller IP

yuyuranium commented:
BTW, I added the vitis project and removed the original controller IP. Also, here is the block diagram:

![image](https://user-images.githubusercontent.com/79467307/169888539-04ca4447-6b70-47cc-a33f-3f30779eb185.png)

japoka410666 reviewed (approved):
Thanks for your hrad working anb testing. It looks great !!
